File: src/org/telegram/messenger/UserConfig.ts
/**
 * Telegram Official UserConfig
 * Replicates org.telegram.messenger.UserConfig.java from DrKLO/Telegram
 * Manages multi-account profiles, activation state, and isolated account keys.
 */

import { NotificationCenter } from './NotificationCenter';

export interface TelegramCurrentUser {
  id: number | string;
  first_name?: string;
  last_name?: string;
  username?: string;
  phone?: string;
  photo?: string;
  bio?: string;
  is_premium?: boolean;
  is_verified?: boolean;
  has_2fa?: boolean;
  two_factor_enabled?: boolean;
}

export class UserConfig {
  public static readonly MAX_ACCOUNT_COUNT = 4;
  public static selectedAccount: number = 0;
  private static instances: UserConfig[] = [];

  public readonly currentAccount: number;
  private currentUser: TelegramCurrentUser | null = null;
  private clientUserId: number | string = 0;
  private isActivated: boolean = false;

  private constructor(account: number) {
    this.currentAccount = account;
    this.loadConfig();
  }

  public static getInstance(account: number = UserConfig.selectedAccount): UserConfig {
    if (account < 0 || account >= UserConfig.MAX_ACCOUNT_COUNT) {
      account = 0;
    }
    if (!UserConfig.instances[account]) {
      UserConfig.instances[account] = new UserConfig(account);
    }
    return UserConfig.instances[account];
  }

  public static getActivatedAccountsCount(): number {
    let count = 0;
    for (let i = 0; i < UserConfig.MAX_ACCOUNT_COUNT; i++) {
      if (UserConfig.getInstance(i).isClientActivated()) {
        count++;
      }
    }
    return count;
  }

  public static getClientUserId(): number | string {
    return UserConfig.getInstance().getClientUserId();
  }

  public static getCurrentUser(): TelegramCurrentUser | null {
    return UserConfig.getInstance().getCurrentUser();
  }

  public static isClientActivated(): boolean {
    return UserConfig.getInstance().isClientActivated();
  }

  public getClientUserId(): number | string {
    if (!this.clientUserId) {
      this.loadConfig();
    }
    return this.clientUserId;
  }

  public getCurrentUser(): TelegramCurrentUser | null {
    if (!this.currentUser) {
      this.loadConfig();
    }
    return this.currentUser;
  }

  public isClientActivated(): boolean {
    return this.isActivated && Boolean(this.clientUserId);
  }

  public isPremium(): boolean {
    return Boolean(this.currentUser?.is_premium);
  }

  public setCurrentUser(user: TelegramCurrentUser | null): void {
    this.currentUser = user;
    if (user && user.id) {
      this.clientUserId = user.id;
      this.isActivated = true;
    } else {
      this.clientUserId = 0;
      this.isActivated = false;
    }
    this.saveConfig();
    NotificationCenter.getInstance(this.currentAccount).postNotificationName(NotificationCenter.mainUserInfoChanged);
  }

  public clearConfig(): void {
    this.currentUser = null;
    this.clientUserId = 0;
    this.isActivated = false;
    try {
      localStorage.removeItem(`tg_user_config_${this.currentAccount}`);
    } catch {}
    NotificationCenter.getInstance(this.currentAccount).postNotificationName(NotificationCenter.appDidLogout);
  }

  public loadConfig(): void {
    try {
      const saved = localStorage.getItem(`tg_user_config_${this.currentAccount}`) || (this.currentAccount === 0 ? localStorage.getItem('tg_user_config') : null);
      if (saved) {
        const parsed = JSON.parse(saved);
        this.currentUser = parsed.user || null;
        this.clientUserId = parsed.clientUserId || (parsed.user?.id || 0);
        this.isActivated = Boolean(this.clientUserId);
      }
    } catch {}
  }

  public saveConfig(): void {
    try {
      localStorage.setItem(
        `tg_user_config_${this.currentAccount}`,
        JSON.stringify({
          clientUserId: this.clientUserId,
          user: this.currentUser,
          currentAccount: this.currentAccount,
        })
      );
    } catch {}
  }
}

File: TMessagesProject/src/main/java/org/telegram/messenger/UserConfig.java
package org.telegram.messenger;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * UserConfig - Replicates DrKLO/Telegram UserConfig.java
 * Manages Multi-Account state (up to MAX_ACCOUNT_COUNT), active user profiles,
 * and account isolation tokens.
 */
public class UserConfig {
    public static final int MAX_ACCOUNT_COUNT = 4;
    public static int selectedAccount = 0;

    private static final UserConfig[] Instance = new UserConfig[MAX_ACCOUNT_COUNT];
    private final int currentAccount;

    public long clientUserId;
    public String currentPhoneNumber;
    public String currentFirstName;
    public String currentLastName;
    public String currentUsername;
    public boolean isPremium;
    public boolean has2fa;
    public boolean registered;

    public static UserConfig getInstance(int account) {
        if (account < 0 || account >= MAX_ACCOUNT_COUNT) {
            account = 0;
        }
        UserConfig local = Instance[account];
        if (local == null) {
            synchronized (UserConfig.class) {
                local = Instance[account];
                if (local == null) {
                    Instance[account] = local = new UserConfig(account);
                }
            }
        }
        return local;
    }

    private UserConfig(int account) {
        currentAccount = account;
        loadConfig();
    }

    public static int getActivatedAccountsCount() {
        int count = 0;
        for (int a = 0; a < MAX_ACCOUNT_COUNT; a++) {
            if (getInstance(a).isClientActivated()) {
                count++;
            }
        }
        return count;
    }

    public boolean isClientActivated() {
        return clientUserId != 0 && registered;
    }

    public void setCurrentUser(long userId, String phone, String firstName, String lastName, String username, boolean premium, boolean twoFa) {
        this.clientUserId = userId;
        this.currentPhoneNumber = phone;
        this.currentFirstName = firstName;
        this.currentLastName = lastName;
        this.currentUsername = username;
        this.isPremium = premium;
        this.has2fa = twoFa;
        this.registered = (userId != 0);
        saveConfig(true);
        NotificationCenter.getInstance(currentAccount).postNotificationName(NotificationCenter.mainUserInfoChanged);
    }

    public void clearConfig() {
        this.clientUserId = 0;
        this.currentPhoneNumber = null;
        this.currentFirstName = null;
        this.currentLastName = null;
        this.currentUsername = null;
        this.isPremium = false;
        this.has2fa = false;
        this.registered = false;
        saveConfig(true);
        NotificationCenter.getInstance(currentAccount).postNotificationName(NotificationCenter.appDidLogout);
    }

    public void loadConfig() {
        Context context = ApplicationLoader.applicationContext;
        if (context == null) return;
        SharedPreferences preferences = context.getSharedPreferences("userconfig_" + currentAccount, Context.MODE_PRIVATE);
        clientUserId = preferences.getLong("clientUserId", 0);
        currentPhoneNumber = preferences.getString("phone", null);
        currentFirstName = preferences.getString("firstName", null);
        currentLastName = preferences.getString("lastName", null);
        currentUsername = preferences.getString("username", null);
        isPremium = preferences.getBoolean("isPremium", false);
        has2fa = preferences.getBoolean("has2fa", false);
        registered = preferences.getBoolean("registered", clientUserId != 0);
    }

    public void saveConfig(boolean notify) {
        Context context = ApplicationLoader.applicationContext;
        if (context == null) return;
        SharedPreferences preferences = context.getSharedPreferences("userconfig_" + currentAccount, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putLong("clientUserId", clientUserId);
        editor.putString("phone", currentPhoneNumber);
        editor.putString("firstName", currentFirstName);
        editor.putString("lastName", currentLastName);
        editor.putString("username", currentUsername);
        editor.putBoolean("isPremium", isPremium);
        editor.putBoolean("has2fa", has2fa);
        editor.putBoolean("registered", registered);
        editor.apply();
    }
}
